hls-plugin-api/src/Ide/Plugin/Error.hs

```diff
@@ -0,0 +1,163 @@
+    -- client.
+    --
+    -- This error will be logged individually with Warning, and will be
+    -- converted into a ParseError response code. It takes a medium precedence
+    --  (2) in being returned as a response to the client.
+  | PluginParseError T.Text
+    -- |PluginDependencyFailed should be thrown when a function that your plugin
+    -- depends on fails. This should only be used when the function fails
+    -- because the files the user is working on is in an invalid state.
+    --
+    -- This error takes the name of the function that failed. Prefer to catch
+    -- this error as close to the source as possible.
+    --
+    -- This error will be logged together with other errors of the same type
+    -- with Info, and will be converted into a ContentModified response. It
+    -- takes a low precedence (3) in being returned as a response to the client.
+  | PluginDependencyFailed T.Text
+    -- |PluginRequestRefused allows your handler to inspect a request before
+    -- rejecting it. In effect it allows your plugin to act make a secondary
+    -- `pluginEnabled` decision after receiving the request. This should only be
+    -- used if the decision to accept the request can not be made in
+    -- `pluginEnabled`.
+    --
+    -- This error will be logged together with other errors of the same type
+    -- with Info. If it's the only response to a request, HLS will respond as if
+    -- no plugins passed the `pluginEnabled` stage.
+  | PluginRequestRefused
+    -- |PluginRuleFailed should be thrown when a Rule your response depends on
+    -- fails.
+    --
+    -- This error takes the name of the Rule that failed.
+    --
+    -- This error will be logged together with other errors of the same type
+    -- with Info, and will be converted into a ContentModified response code. It
+    -- takes a low precedence (3) in being returned as a response to the client.
+  | PluginRuleFailed T.Text
+    -- |PluginStaleResolve should be thrown when your resolve request is
+    -- provided with data it can no longer resolve.
+    --
+    -- This error will be logged individually with Info, and will be converted
+    -- into a ContentModified response. It takes a low precedence (3) in being
+    -- returned as a response to the client.
+  | PluginStaleResolve
+
+instance Pretty PluginError where
+    pretty = \case
+      PluginInternalError msg -> "Internal Error:" <+> pretty msg
+      PluginStaleResolve -> "Stale Resolve"
+      PluginRuleFailed rule       -> "Rule Failed:" <+> pretty rule
+      PluginInvalidParams text -> "Invalid Params:" <+> pretty text
+      PluginParseError text -> "Parse Error:" <+> pretty text
+      PluginInvalidRequest text -> "Invalid Request:" <+> pretty text
+      PluginDependencyFailed text -> "Dependency Failed:" <+> pretty text
+      PluginRequestRefused -> "Request Refused"
+
+handleMaybe :: Monad m => e -> Maybe b -> ExceptT e m b
```

Comment:
I have mixed feelings about these. They are quite common, but also I think people can write them inline. This one in particular really just saves people a case expression, and I would usually rather read the case expression!